### ISSUE TITLE
Unable to provide macro definition if CLANG_ASSISTED_PARSING is enabled

### ISSUE THREAD

**Issue by Bagira80** [doxygen/doxygen#7919]
I built Doxygen with support for Clang 11 enabled and am able to to use Doxygen with the clang-parser when setting `CLANG_ASSISTED_PARSING` to `YES`.  
However, I need to provide a preprocessor definition `DOXYGEN` when parsing via Doxygen/clang-parser, because my code has some special handling like the following:

```c++
#ifdef DOXYGEN
// do something related to Doxygen
#else
// do something related to normal compilation
#endif
```

If `CLANG_ASSISTED_PARSING` is set to `NO` I can provide that macro using Doxygen's `PREDEFINED` setting.  
However, if `CLANG_ASSISTED_PARSING` is set to `YES` that setting is ignored.

I thought I could use `CLANG_OPTIONS` and give it a value `-DDOXYGEN=1` (in order to use Clang's `-D` flag) but that seems to have no effect, too.

*What am I doing wrong?*

I am using a custom-built Doxygen executable, because I wanted to use the latest release and link against clang-parser from Clang 11. It's version is `1.8.18 (a1b07ad0e92e4526c9ba1711d39f06b58c2a7459*)`  
*Note:* I had to apply the patch from [here](https://github.com/doxygen/doxygen/issues/7883#issuecomment-663599967) in order to build with support for clang-parser from Clang 11.

**Comment by albert-github** [comment]
I tried it  it with the current master (on Cygwin and with a statically linked executable) and this looks OK to me.
My doxygen configuration file (`doxygen -x`):
```
QUIET                  = YES
CLANG_ASSISTED_PARSING = YES
PREDEFINED             = DOXYGEN
```
my source:
```
/// \file

#ifdef DOXYGEN
/// the doxygen version
void doxy_fie(void);
#else
/// the normal version
void normal_fie(void);
#endif
```

running `doxygen` gives:
```
00001 /// \file
00002
00003
00004 /// the doxygen version
00005 void doxy_fie(void);
00006
00007
00008
00009
00010
```
i.e. the preprocessor output and this looks OK to me. Also the resulting documentation looks OK to me.

- What happens when you run the above code with your executable?
- What happens when you create an executable from the current master (already contains a fix for the dynamically linked clang  parser)?

**Comment by Bagira80** [comment]
OK, I must correct myself. You are right.

Using your example and trying to simplify my (production) example showed that `PREDEFINED` is **not** ignored, even in Clang mode.  
But it seems, trying to add a macro definition through `CLANG_OPTIONS` (e.g. `CLANG_OPTIONS = -DDOXYGEN`) does not have any impact and gets ignored. (It even seems that `CLANG_OPTIONS` is ignored entirely.)

---

I was originally mislead to think that the macro from `PREDEFINED` did not get defined because the Clang parser complains about a `use of undeclared identifier 'std'`.

In order to fix that error, I then tried to conditionally include the required C++ standard header like so:
```c++
#ifdef DOXYGEN
#include <type_traits>
#endif
```
And this did not help.

However, I now realized that omitting the `#ifdef DOXYGEN` and unconditionally including the standards header gets rid of the error.  
This is inconvenient (as normal code could very well have conditional includes) and should probably be logged as another Doxygen issue.

I am not sure if this behavior has something to do with the fact that I am using the `CLANG_DATABASE_PATH` variable or that the Clang database (`compile_commands.json`) file was generated by CMake using GCC, and Doxygen uses it for the Clang parser?  
Using the Clang database file generated by CMake using Clang 11 (with libc++) brings even more weird errors, like `fatal error: 'type_traits' file not found`.

Adding the macro-definition to the compiler command in `compile_commands.json` defines the macro while parsing and the conditional include succeeds.

**Comment by albert-github** [comment]
Indeed it should be possible to exclude included files by means of `#ifdef` constructs.

I don't see what is, fundamentally, different from my example and the `#include` example but when I run doxygen with the preprocessor output (`doxygen -d preprocessor`) I see that the include is gone, but also a warning:
```
#include type_traits: not found! skipping...
```

Maybe you should add the path to `type_traits` to the doxygen setting `INCLUDE_PATH`

**Comment by doxygen** [comment]
@Bagira80 and @albert-github Note that option `CLANG_OPTIONS` is to specify additional flags that make the clang parser understand your code (e.g. compile it without errors). This parser is used along-site doxygen's built in parser. The clang parser requires code to compile, while doxygen which doesn't mind if the code is incomplete and you might even use the preprocessor to make doxygen ignore certain parts that are needed when compiling the code (like the inclusion of `type_traits`).

**Comment by Bagira80** [comment]
> Note that option `CLANG_OPTIONS` is to specify additional flags that make the clang parser understand your code (e.g. compile it without errors).

Are these all the normal flags the Clang compiler can take?

My original code compiles fine with Clang 11 and libc++ and GCC 9 (with libstdc++). The generated `compile_commands.json` contains the compiler-commands for all source-files (but not header-files!) with all the required options for the chosen compiler to compile them fine.

So, using the `compile_commands.json` file generated with Clang 11 I would have expected that Doxygen's clang parser can parse the code fine without errors or additional help. But it fails miserably because it cannot even find the C++ standard headers (provided by libc++).  
Adding the directory which contains these to the include search path via `CLANG_OPTIONS` or `INCLUDE_PATH` does not help.

Interestingly, using the `compile_commands.json` file generated with GCC 9 pleases Doxygen's clang parser mostly. However, it still chokes on parsing headers that are not fully self-contained (e.g. are missing an `#include <type_traits>` that is normally provided by the source-file that included that header).  
Adding the directory which contains these standard headers to the include search path does also not help, whether they are from libc++ or libstdc++.

For such non fully self-contained headers I thought to conditionally include the missing standard headers guarded by a macro (`DOXYGEN`) that is only defined when running Doxygen. And that did not succeed (which led to this issue ticket).

---

> This parser is used along-site doxygen's built in parser.

I don't understand that.  
Do you mean, that if I set `CLANG_ASSISTED_PARSING` to `YES` Doxygen parses the code as always but that additionally the clang parser parses it, too? I don't believe that is what you meant.

---

> The clang parser requires code to compile, ...

The code compiles fine if clang parser is only starting at the source files (which provide missing includes for the non self-contained headers they include) and not at the headers directly.

> ... while doxygen which doesn't mind if the code is incomplete and you might even use the preprocessor to make doxygen ignore certain parts that are needed when compiling the code (like the inclusion of `type_traits`).

In my case I wanted to use the preprocessor to include extra code. :smile:

---

@doxygen Maybe you can shortly explain the purpose of the clang parser and its interaction with Doxygen (and what original functionality of Doxygen it might replace) as well as how the clang parsers really uses the `compile_commands.json` file?

Maybe I just made some mistake when compiling Doxygen with Clang 11 support? I might need to try to recompile it with HEAD of `master` branch as @albert-github already suggested.

**Comment by Bagira80** [comment]
OK, with https://github.com/doxygen/doxygen/pull/7922 I am now able to compile (the HEAD of `master` branch of) Doxygen with GCC or Clang with or without static or shared clang parser support.

Regardless of the way I build Doxygen, as long as I compiled in clang parser support, the result is always the same:

Header files which are *not self-contained* and conditionally include required headers if a specific preprocessor macro is defined by the parser, still result in the clang parser complaining about unknown identifiers, regardless of how I define that macro.  
Only hard-coding its definition into the source-file (or the `compile_commands.json` file) works. But that cancels its purposes entirely.

**Comment by Bagira80** [comment]
I must correct my last comment:

#7922 fixes compiling Doxygen with Clang and also fixes compiling Doxygen with statically linked `libclang-cpp.so` (aka `libclangTooling.a`). But this has nothing to do with the problem described in this issue.

This issue is actually a *user problem* (or an *unclear documentation problem*).

---

As long as I have a `doxygen` executable which has clang parser support built-in, I can fix the described problem.

However, I have to provide the required preprocessor definition `DOXYGEN` both, via settings `PREDEFINED` and `CLANG_OPTIONS`!

Example Doxyfile:
```
CLANG_ASSISTED_PARSING = YES
CLANG_DATABASE_PATH    = <path-to-directory-containing-compile_commands.json>
CLANG_OPTIONS          = -DDOXYGEN=1
PREDEFINED             = DOXYGEN=1
```
Example source:
```c++
...
#ifdef DOXYGEN
#include <type_traits>  // Required to prevent clang parser error about unknown identifer 'std'.

/// Only in the documentation if preprocessor definition 'DOXYGEN' is defined in setting 'PREDEFINED'.
struct MyStruct{};
#endif
...

using TrueType = std::true_type;  // No clang parser error if preprocessor definition 'DOXYGEN' is defined in setting 'CLANG_OPTIONS'.


```

And I now also understand the different meanings:

* The error message I got from the clang-parser (`error: use of undeclared identifier 'std' [clang]`) occurs if the preprocessor definition `DOXYGEN` is missing in the `CLANG_OPTIONS` setting.
* The documentation for `MyStruct` is not generated if the preprocessor definition `DOXYGEN` is missing in the `PREDEFINED` settings.

---

**In a nutshell: One should always provide required preprocessor definitions through both doxygen settings, `PREDEFINED` and `CLANG_OPTIONS`!**

**Comment by Bagira80** [comment]
I still do not understand why Doxygen has to parse the code itself when clang parser already did so, but at least it works, now.  
Therefore, I would say this issue can be closed.

However, I would recommend to document this behavior in Doxygen's manual. Probably, the help for `PREDEFINED` and `CLANG_OPTIONS` should mention that explicitly. It is not intuitive.

**Comment by doxygen** [comment]
@Bagira80 Doxygen uses libclang only to do the syntax highlighting and cross-referencing. For the so called "outline parsing", e.g. figuring what symbols are in the code that can be documented, libclang is not used. This was a design choice. The main advantage is that you still get output in case something is not part of a translation unit. Think of a library providing some header file where the implementation is in the header; with only libclang such header will not be parsed if the scope of doxygen is just the library and does not include the client using it. Another reason not to use libclang for the outline parsing is the preprocessor. libclang provides symbols after preprocessing, whereas doxygen provides more control over the processing step itself, also allowing for instance macros to be documented, and expanded in a certain way.

**Comment by Bagira80** [comment]
Thanks for the clarification.  
Is that rationale somewhere found in Doxygen's manual? I think it belongs there.

Regarding the definition of preprocessor symbols in `PREDEFINED` and `CLANG_PARSER`: I still recommend to document that behavior in Doxygen's manual. As I already wrote, it is not intuitive. (If it were I would not have created this issue in the first place. :grin:)
